### ISSUE TITLE
Some users don't have permission to view query with a subquery, but can view the subquery directly

### ISSUE THREAD

**Issue by hobailey** [metabase/metabase#12354]
**Describe the bug**
Some users (but not all) can not access a question which references another question (via a subquery).
![image](https://user-images.githubusercontent.com/1564439/79759483-a476f500-8316-11ea-8538-7da03021a9e5.png)

If the user tries to access the reference question directly (question ID 1230 in this case), they have access to view it without any error at all.

Note that only _some_ users are affected, not all. I think it might be related to the version that moved some questions to the "Migrated questions" collection because users added to Metabase _since_ then do not experience the same issue, even though they have identical rights.

I also tried putting the referenced question in the same collection as the main question, but that made no difference to the above error for the users in question.

**To Reproduce**
1. Create a question
2. Create another question and reference the first one
3. Check that all users who can access question 1 can also access question 2 (some may see the error described)

**Expected behavior**
The user to be able to access referenced questions

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.5+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.5",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.5+10",
    "os.name": "Linux",
    "os.version": "4.4.0-1095-aws",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "bigquery",
      "athena",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "9.6.11"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-04-10",
      "tag": "v0.35.2",
      "branch": "release-0.35.x",
      "hash": "f3b2857"
    },
    "settings": {
      "report-timezone": "Europe/London"
    }
  }
}
```

- Your databases: Postgres
- Metabase version: 0.35.2
- Metabase hosting environment: Docker
- Metabase internal database: Postgres

**Severity**
This is blocking us being able to use the new "reference question" feature

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by flamber** [comment]
Just adding the error as text, so it can be searched: `The sub-query from referenced question #1230 failed with the following error: You do not have permissions to run this query`

Reproduced with a user with view-only to "All Users" collection, no data-access, where both questions (Sample Dataset) are placed in "Our Analytics". The user can view the first question, but gets an error on the second question, which is referencing the first question.

**Comment by hobailey** [comment]
Thanks @flamber - did you find a workaround for the time being by any chance?

**Comment by flamber** [comment]
Nope, I'm sure this is a backend protection layer that is too strict. Rather have too much security than lack of it ;-)
But I'm fairly sure it's not related to "Migrated questions", which I believe was 0.30 - I'm glad about that, since it would take a lot time to test and migrate such a problem if that was the case.

**Comment by karthik-vv** [comment]
Hi facing similar issue.
I have created a separate collection where i store commonly used questions "COMMON". and i have given highest access (edit) to a group (group of users) 'GROUP1". 

1. Question in one of there personal collection which referees ({{#1234}})to a question in "COMMON" collection they get error "sub-query from referenced question failed".
2. Users of  "GROUP1"  also have access to different collection "COL_G1" now a question saved in "COL_G1" is referring to a question in "COMMON", they get error "sub-query from referenced question failed".

**Comment by hobailey** [comment]
@flamber any ideas roughly when this might be released by any chance? It renders the feature totally unusable atm :-(

**Comment by flamber** [comment]
@hobailey It has been given priority, but I cannot give you a date.

**Comment by dacort** [comment]
@camsaul Going to reopen this - unsure if I'm doing something wrong, but the repro I wrote in https://github.com/metabase/metabase/pull/12525 seems to indicate this still exists.

**Comment by flamber** [comment]
It looks like it's still expecting Native permissions - tested on commit f58c7043302c6ee19576fdc14ddea643e8800e23
<details><summary>Error log</summary>

```
05-22 19:19:37 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2020-05-22T19:19:37.103507+02:00[Europe/Copenhagen]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Error determining value for parameter",
   :stacktrace
   ["--> driver.common.parameters.values$eval58646$value_for_tag__58651$fn__58652.invoke(values.clj:273)"
    "driver.common.parameters.values$eval58646$value_for_tag__58651.invoke(values.clj:262)"
    "driver.common.parameters.values$eval58673$query__GT_params_map__58678$fn__58682$iter__58684__58688$fn__58689.invoke(values.clj:287)"
    "driver.common.parameters.values$eval58673$query__GT_params_map__58678$fn__58682.invoke(values.clj:286)"
    "driver.common.parameters.values$eval58673$query__GT_params_map__58678.invoke(values.clj:277)"
    "driver.sql$eval59355$fn__59357.invoke(sql.clj:45)"
    "query_processor.middleware.parameters.native$expand_inner.invokeStatic(native.clj:39)"
    "query_processor.middleware.parameters.native$expand_inner.invoke(native.clj:30)"
    "query_processor.middleware.parameters$expand_one.invokeStatic(parameters.clj:52)"
    "query_processor.middleware.parameters$expand_one.invoke(parameters.clj:43)"
    "query_processor.middleware.parameters$expand_all$replace_54246__54247.invoke(parameters.clj:61)"
    "mbql.util.match$replace_in_collection$iter__31483__31487$fn__31488.invoke(match.clj:132)"
    "mbql.util.match$replace_in_collection.invokeStatic(match.clj:131)"
    "mbql.util.match$replace_in_collection.invoke(match.clj:126)"
    "query_processor.middleware.parameters$expand_all$replace_54246__54247.invoke(parameters.clj:61)"
    "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:61)"
    "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:55)"
    "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:58)"
    "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:55)"
    "query_processor.middleware.parameters$expand_parameters.invokeStatic(parameters.clj:79)"
    "query_processor.middleware.parameters$expand_parameters.invoke(parameters.clj:75)"
    "query_processor.middleware.parameters$eval54262$substitute_parameters_STAR___54267$fn__54268.invoke(parameters.clj:84)"
    "query_processor.middleware.parameters$eval54262$substitute_parameters_STAR___54267.invoke(parameters.clj:81)"
    "query_processor.middleware.parameters$substitute_parameters$fn__54283.invoke(parameters.clj:99)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__52541.invoke(resolve_referenced.clj:80)"
    "query_processor.middleware.expand_macros$expand_macros$fn__53204.invoke(expand_macros.clj:158)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__51283.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__57988.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54549$fn__54553.invoke(resolve_database_and_driver.clj:33)"
    "driver$do_with_driver.invokeStatic(driver.clj:61)"
    "driver$do_with_driver.invoke(driver.clj:57)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54549.invoke(resolve_database_and_driver.clj:27)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__53430.invoke(fetch_source_query.clj:267)"
    "query_processor.middleware.store$initialize_store$fn__58005$fn__58006.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
    "query_processor.store$do_with_store.invoke(store.clj:40)"
    "query_processor.middleware.store$initialize_store$fn__58005.invoke(store.clj:10)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__52384.invoke(cache.clj:209)"
    "query_processor.middleware.validate$validate_query$fn__58022.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__53578.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__51131.invoke(add_rows_truncated.clj:36)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__57967.invoke(results_metadata.clj:147)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__52803.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__54398.invoke(process_userland_query.clj:136)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__52732.invoke(catch_exceptions.clj:174)"
    "query_processor.reducible$async_qp$qp_STAR___46866$thunk__46867.invoke(reducible.clj:101)"
    "query_processor.reducible$async_qp$qp_STAR___46866$fn__46869.invoke(reducible.clj:106)"],
   :ex-data {:tag {:id "5466a2a1-85d2-140c-58ec-5ac09da88f67", :name "#312", :display-name "#312", :type :card, :card-id 312}}}
  {:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Error building query parameter map",
   :stacktrace
   ["--> driver.common.parameters.values$eval58673$query__GT_params_map__58678$fn__58682.invoke(values.clj:295)"
    "driver.common.parameters.values$eval58673$query__GT_params_map__58678.invoke(values.clj:277)"
    "driver.sql$eval59355$fn__59357.invoke(sql.clj:45)"
    "query_processor.middleware.parameters.native$expand_inner.invokeStatic(native.clj:39)"
    "query_processor.middleware.parameters.native$expand_inner.invoke(native.clj:30)"
    "query_processor.middleware.parameters$expand_one.invokeStatic(parameters.clj:52)"
    "query_processor.middleware.parameters$expand_one.invoke(parameters.clj:43)"
    "query_processor.middleware.parameters$expand_all$replace_54246__54247.invoke(parameters.clj:61)"
    "mbql.util.match$replace_in_collection$iter__31483__31487$fn__31488.invoke(match.clj:132)"
    "mbql.util.match$replace_in_collection.invokeStatic(match.clj:131)"
    "mbql.util.match$replace_in_collection.invoke(match.clj:126)"
    "query_processor.middleware.parameters$expand_all$replace_54246__54247.invoke(parameters.clj:61)"
    "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:61)"
    "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:55)"
    "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:58)"
    "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:55)"
    "query_processor.middleware.parameters$expand_parameters.invokeStatic(parameters.clj:79)"
    "query_processor.middleware.parameters$expand_parameters.invoke(parameters.clj:75)"
    "query_processor.middleware.parameters$eval54262$substitute_parameters_STAR___54267$fn__54268.invoke(parameters.clj:84)"
    "query_processor.middleware.parameters$eval54262$substitute_parameters_STAR___54267.invoke(parameters.clj:81)"
    "query_processor.middleware.parameters$substitute_parameters$fn__54283.invoke(parameters.clj:99)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__52541.invoke(resolve_referenced.clj:80)"
    "query_processor.middleware.expand_macros$expand_macros$fn__53204.invoke(expand_macros.clj:158)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__51283.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__57988.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54549$fn__54553.invoke(resolve_database_and_driver.clj:33)"
    "driver$do_with_driver.invokeStatic(driver.clj:61)"
    "driver$do_with_driver.invoke(driver.clj:57)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54549.invoke(resolve_database_and_driver.clj:27)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__53430.invoke(fetch_source_query.clj:267)"
    "query_processor.middleware.store$initialize_store$fn__58005$fn__58006.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
    "query_processor.store$do_with_store.invoke(store.clj:40)"
    "query_processor.middleware.store$initialize_store$fn__58005.invoke(store.clj:10)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__52384.invoke(cache.clj:209)"
    "query_processor.middleware.validate$validate_query$fn__58022.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__53578.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__51131.invoke(add_rows_truncated.clj:36)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__57967.invoke(results_metadata.clj:147)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__52803.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__54398.invoke(process_userland_query.clj:136)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__52732.invoke(catch_exceptions.clj:174)"
    "query_processor.reducible$async_qp$qp_STAR___46866$thunk__46867.invoke(reducible.clj:101)"
    "query_processor.reducible$async_qp$qp_STAR___46866$fn__46869.invoke(reducible.clj:106)"],
   :error_type :invalid-parameter,
   :ex-data {:type :invalid-parameter, :tags {"#312" {:id "5466a2a1-85d2-140c-58ec-5ac09da88f67", :name "#312", :display-name "#312", :type :card, :card-id 312}}, :params nil}}],
 :error_type :missing-required-permissions,
 :json_query
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :native,
  :middleware nil,
  :native {:query "SELECT * FROM {{#312}}", :template-tags {"#312" {:id "5466a2a1-85d2-140c-58ec-5ac09da88f67", :name "#312", :display-name "#312", :type :card, :card-id 312}}},
  :database 4,
  :parameters [],
  :async? true,
  :cache-ttl nil},
 :status :failed,
 :class clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> driver.common.parameters.values$eval58406$card_query_for_tag__58411$fn__58412.invoke(values.clj:137)"
  "driver.common.parameters.values$eval58406$card_query_for_tag__58411.invoke(values.clj:127)"
  "driver.common.parameters.values$eval58646$value_for_tag__58651$fn__58652.invoke(values.clj:269)"
  "driver.common.parameters.values$eval58646$value_for_tag__58651.invoke(values.clj:262)"
  "driver.common.parameters.values$eval58673$query__GT_params_map__58678$fn__58682$iter__58684__58688$fn__58689.invoke(values.clj:287)"
  "driver.common.parameters.values$eval58673$query__GT_params_map__58678$fn__58682.invoke(values.clj:286)"
  "driver.common.parameters.values$eval58673$query__GT_params_map__58678.invoke(values.clj:277)"
  "driver.sql$eval59355$fn__59357.invoke(sql.clj:45)"
  "query_processor.middleware.parameters.native$expand_inner.invokeStatic(native.clj:39)"
  "query_processor.middleware.parameters.native$expand_inner.invoke(native.clj:30)"
  "query_processor.middleware.parameters$expand_one.invokeStatic(parameters.clj:52)"
  "query_processor.middleware.parameters$expand_one.invoke(parameters.clj:43)"
  "query_processor.middleware.parameters$expand_all$replace_54246__54247.invoke(parameters.clj:61)"
  "mbql.util.match$replace_in_collection$iter__31483__31487$fn__31488.invoke(match.clj:132)"
  "mbql.util.match$replace_in_collection.invokeStatic(match.clj:131)"
  "mbql.util.match$replace_in_collection.invoke(match.clj:126)"
  "query_processor.middleware.parameters$expand_all$replace_54246__54247.invoke(parameters.clj:61)"
  "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:61)"
  "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:55)"
  "query_processor.middleware.parameters$expand_all.invokeStatic(parameters.clj:58)"
  "query_processor.middleware.parameters$expand_all.invoke(parameters.clj:55)"
  "query_processor.middleware.parameters$expand_parameters.invokeStatic(parameters.clj:79)"
  "query_processor.middleware.parameters$expand_parameters.invoke(parameters.clj:75)"
  "query_processor.middleware.parameters$eval54262$substitute_parameters_STAR___54267$fn__54268.invoke(parameters.clj:84)"
  "query_processor.middleware.parameters$eval54262$substitute_parameters_STAR___54267.invoke(parameters.clj:81)"
  "query_processor.middleware.parameters$substitute_parameters$fn__54283.invoke(parameters.clj:99)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__52541.invoke(resolve_referenced.clj:80)"
  "query_processor.middleware.expand_macros$expand_macros$fn__53204.invoke(expand_macros.clj:158)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__51283.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__57988.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54549$fn__54553.invoke(resolve_database_and_driver.clj:33)"
  "driver$do_with_driver.invokeStatic(driver.clj:61)"
  "driver$do_with_driver.invoke(driver.clj:57)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54549.invoke(resolve_database_and_driver.clj:27)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__53430.invoke(fetch_source_query.clj:267)"
  "query_processor.middleware.store$initialize_store$fn__58005$fn__58006.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__58005.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__52384.invoke(cache.clj:209)"
  "query_processor.middleware.validate$validate_query$fn__58022.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__53578.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__51131.invoke(add_rows_truncated.clj:36)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__57967.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__52803.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__54398.invoke(process_userland_query.clj:136)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__52732.invoke(catch_exceptions.clj:174)"
  "query_processor.reducible$async_qp$qp_STAR___46866$thunk__46867.invoke(reducible.clj:101)"
  "query_processor.reducible$async_qp$qp_STAR___46866$fn__46869.invoke(reducible.clj:106)"],
 :context :question,
 :error "The sub-query from referenced question #312 failed with the following error: You do not have permissions to run this query.",
 :row_count 0,
 :running_time 0,
 :ex-data
 {:type :missing-required-permissions,
  :required-permissions "/db/4/native/",
  :actual-permissions #{"/collection/4/read/" "/collection/root/read/" "/collection/3/" "/collection/5/read/" "/collection/2/read/"},
  :permissions-error? true,
  :card-query-error? true,
  :card-id 312,
  :tag {:id "5466a2a1-85d2-140c-58ec-5ac09da88f67", :name "#312", :display-name "#312", :type :card, :card-id 312}},
 :data {:rows [], :cols []}}
```

</details>

**Comment by camsaul** [comment]
you still have to have permissions for the cards in question i.e. for the collection(s) those cards belong to. It's tested here https://github.com/metabase/metabase/pull/12505/files#diff-0a61c2620bf7c68ce11abcb9c78be78dR494-R535

When those conditions are satisfied it was working for me in the frontend.

**Comment by dacort** [comment]
OK, let me validate I've got the permissions set up properly in the test.

**Comment by dacort** [comment]
@camsaul I'm using the `nodata` user in the Cypress test, which has access to the `Our Analytics` collection and still getting the error. Is the issue the root collection?

![image](https://user-images.githubusercontent.com/1512/82714979-fa0c2d80-9c45-11ea-9314-9238f397b196.png)

![image](https://user-images.githubusercontent.com/1512/82715035-2d4ebc80-9c46-11ea-86da-9fc885ec2947.png)

![image](https://user-images.githubusercontent.com/1512/82715060-47889a80-9c46-11ea-8028-910a8103523f.png)

**Comment by camsaul** [comment]
After looking at the tests in #12525, this is indeed fixed and working correctly, the tests are just testing the wrong thing and behavior is expected

**Comment by camsaul** [comment]
So it seems that the e2e tests in #12525 are testing a case that's not addressed in the backend tests (SQL queries nesting other SQL queries using the `{{#id}}` syntax) so I'm going to make sure that's covered before this issue can be considered fully resolved.

**Comment by camsaul** [comment]
Fixed in #12525

**Comment by hobailey** [comment]
Fix works well and has made a huge difference to us - thanks to everyone who worked on this :-)
